opt cmdDel: ignore pod's netns has gone

cyclinder opened the pull request:
Please ensure your pull request adheres to the following guidelines:

- [ ] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [ ] All code is covered by unit and/or runtime tests where feasible.
- [ ] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [ ] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [ ] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [ ] Provide a title or release-note blurb suitable for the release notes.
- [ ] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)
- [ ] Thanks for contributing!

<!-- Description of change -->

opt cmdDel: ignore pod's netns has gone

```release-note
<!-- Enter the release note text here if needed or remove this section! -->
```

squeed reviewed (changes requested):
This is a good idea, but I think we can do one better.

Rather than trying to interpret every error message, the code should do the following:

1. Submit the deletion request to the server. If that fails, return error
2. Try and open the netns and delete the interface. If any of that fails, give up and return no error.

Make sense? Thanks!

squeed commented on pkg/netns/netns.go:
```diff
@@ -23,7 +22,7 @@ func RemoveIfFromNetNSIfExists(netNS ns.NetNS, ifName string) error {
 		l, err := netlink.LinkByName(ifName)
 		if err != nil {
-			if errors.As(err, &netlink.LinkNotFoundError{}) {
+			if _, ok := err.(netlink.LinkNotFoundError); ok {
```
why did you change the `errors.As`? Was it not working as intended?

cyclinder replied:
ah, I don't remember it, but you're right. updated now.

cyclinder committed:
opt cmdDel: ignore pod's netns has gone

If pod's netns has gone, we should just return directly rather than return an error.

Signed-off-by: qifeng guo <qifeng.guo@daocloud.io>

ti-mo commented on plugins/cilium-cni/main.go:
```diff
@@ -438,6 +438,10 @@ func cmdAdd(args *skel.CmdArgs) (err error) {
 
 	netNs, err := ns.GetNS(args.Netns)
 	if err != nil {
+		if errors.As(err, &ns.NSPathNotExistErr{}) {
```
Let me get this straight: if we don't manage to open the netns pin from nsfs, we return from `cmdAdd` successfully as if a CNI ADD was successfully completed? I don't think that's a good idea, this signals the container runtime that everything's peachy. Which problem does this patch solve, exactly?

cyclinder replied:
Sorry, Ignore this please, this is my mistake.

ti-mo reviewed (changes requested):
> Yes, I think there were still processes running in it, But I used `ip netns list` to check if the network namespace exists, and I don't find it indeed. I don't know which process is still using this network namespace.

Namespaces are reclaimed when the last process using them exits. That's the only condition. I'm not sure what you were trying to find with `ip netns list` but that command scans `/var/run/ns` for netns handles, which the majority of modern container runtimes do not write to. The canonical way of enumerating all namespace handles on the system is by iterating `/proc/*/ns` in the root namespace.

>> Network namespace management is not the responsibility of the CNI plugin. If the namespace provided for CNI DEL does not exist, the CNI spec is quite clear that the plugin should return success.

> So I think we should delete network namespace resources as much as possible, rather than relying entirely on network namespace to be removed. Of course, we should also ignore some non-existent errors, and return meaningful errors,make the runtime retry the deletion.

Which network namespace resources do you mean here? And how does the proposed patch accomplish this? Cilium isn't going to directly kill processes, so it has zero impact on the lifecycle of a netns managed by the container runtime (short of returning an unsuccessful CNI ADD, of course).
